File: android/app/src/main/java/com/hawkeynl/jarvis/security/ModelAuthorization.kt
package com.hawkeynl.jarvis.security

import java.util.UUID

/** Memory-only model authorization; never stores credentials or signed requests. */
class ModelAuthorization {
    private var generation = UUID.randomUUID()
    private data class Grant(val context: String, val monotonic: Long, val wall: Long)
    private var grant: Grant? = null

    @Synchronized fun ticket(): UUID = generation
    @Synchronized fun accepts(ticket: UUID): Boolean = generation == ticket
    @Synchronized fun invalidate() { generation = UUID.randomUUID(); grant = null }
    @Synchronized fun valid(context: String, ticket: UUID, monotonic: Long, wall: Long): Boolean {
        val current = grant ?: return false
        return generation == ticket && current.context == context &&
            monotonic - current.monotonic in 0 until 300_000L && wall - current.wall in 0 until 300_000L
    }
    @Synchronized fun remember(context: String, ticket: UUID, monotonic: Long, wall: Long): Boolean {
        if (generation != ticket) return false
        grant = Grant(context, monotonic, wall)
        return true
    }
}

File: ios/Jarvis/Services/ModelAuthorization.swift
import Foundation

/// Native, memory-only lease. Lock invalidation is synchronous, including while
/// AuthService is suspended in the OS authentication dialog.
final class ModelAuthorization: @unchecked Sendable {
    private let lock = NSLock()
    private var generation = UUID()
    private var grant: (context: String, uptime: TimeInterval, wall: TimeInterval)?

    func ticket() -> UUID { lock.withLock { generation } }
    func accepts(_ ticket: UUID) -> Bool { lock.withLock { generation == ticket } }
    func invalidate() { lock.withLock { generation = UUID(); grant = nil } }

    func valid(_ context: String, ticket: UUID,
               uptime: TimeInterval = ProcessInfo.processInfo.systemUptime,
               wall: TimeInterval = Date().timeIntervalSince1970) -> Bool {
        lock.withLock {
            guard generation == ticket, let grant, grant.context == context else { return false }
            return (0..<300).contains(uptime - grant.uptime) && (0..<300).contains(wall - grant.wall)
        }
    }

    @discardableResult
    func remember(_ context: String, ticket: UUID,
                  uptime: TimeInterval = ProcessInfo.processInfo.systemUptime,
                  wall: TimeInterval = Date().timeIntervalSince1970) -> Bool {
        lock.withLock {
            guard generation == ticket else { return false }
            grant = (context, uptime, wall)
            return true
        }
    }
}
